runtime/integration-tests/src/generic/envs/runtime_env.rs

```diff
@@ -0,0 +1,200 @@
+
+			let extrinsic =
+				<T::Block as Block>::Extrinsic::new(runtime_call, Some(multi_address)).unwrap();
+
+			self.nonce += 1;
+
+			T::apply_extrinsic(extrinsic).unwrap()
+		})
+	}
+
+	fn pass(&mut self, blocks: Blocks<T>) {
+		self.ext.borrow_mut().execute_with(|| {
+			let next = frame_system::Pallet::<T>::block_number() + 1;
+
+			let end_block = match blocks {
+				Blocks::ByNumber(n) => next + n,
+				Blocks::BySeconds(secs) => {
+					let blocks = secs / pallet_aura::Pallet::<T>::slot_duration();
+					if blocks % pallet_aura::Pallet::<T>::slot_duration() != 0 {
+						blocks as BlockNumber + 1
+					} else {
+						blocks as BlockNumber
+					}
+				}
+				Blocks::UntilEvent { limit, .. } => limit,
+			};
+
+			for i in next..end_block {
+				T::finalize_block();
+				Self::prepare_block(i);
+
+				if let Blocks::UntilEvent { event, .. } = blocks.clone() {
+					let event: T::RuntimeEventExt = event.into();
+					if frame_system::Pallet::<T>::events()
+						.into_iter()
+						.find(|record| record.event == event)
+						.is_some()
+					{
+						break;
+					}
+				}
+			}
+		})
+	}
+
+	fn state_mut<R>(&mut self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(f)
+	}
+
+	fn state<R>(&self, f: impl FnOnce() -> R) -> R {
+		self.ext.borrow_mut().execute_with(|| {
+			transactional::with_transaction(|| {
+				// We revert all changes done by the closure to offer an inmutable state method
+				TransactionOutcome::Rollback::<Result<R, DispatchError>>(Ok(f()))
+			})
+			.unwrap()
+		})
+	}
+}
+
+impl<T: Runtime> RuntimeEnv<T> {
+	fn prepare_block(i: BlockNumber) {
+		let slot = Slot::from(i as u64);
+		let digest = Digest {
+			logs: vec![DigestItem::PreRuntime(AURA_ENGINE_ID, slot.encode())],
+		};
+
+		let header = Header {
+			number: i,
+			digest,
+			state_root: H256::default(),
+			extrinsics_root: H256::default(),
+			parent_hash: H256::default(),
+		};
+
+		T::initialize_block(&header);
+
+		let timestamp = i as u64 * pallet_aura::Pallet::<T>::slot_duration();
+		let inherent_extrinsics = vec![
+			Extrinsic::new(Self::cumulus_inherent(i), None).unwrap(),
+			Extrinsic::new(Self::timestamp_inherent(timestamp), None).unwrap(),
+		];
+
+		for extrinsic in inherent_extrinsics {
+			T::apply_extrinsic(extrinsic).unwrap().unwrap();
+		}
+	}
+
+	fn cumulus_inherent(i: BlockNumber) -> T::RuntimeCall {
+		let mut inherent_data = InherentData::default();
+
+		let sproof_builder = RelayStateSproofBuilder::default();
+		let (relay_parent_storage_root, relay_chain_state) =
+			sproof_builder.into_state_root_and_proof();
+
+		let cumulus_inherent = ParachainInherentData {
+			validation_data: PersistedValidationData {
+				parent_head: vec![].into(),
+				relay_parent_number: i,
+				max_pov_size: Default::default(),
+				relay_parent_storage_root,
+			},
+			relay_chain_state,
+			downward_messages: Default::default(),
+			horizontal_messages: Default::default(),
```

Comment:
We could inject XCM messages here. That would be great.

Reply:
@cdamian

Reply:
This seems like a standalone env - no relay nor other parachains being involved. So, even if we populate the `horizontal_messages`, there's no receiving end for it afaict.

Reply:
Could we mock the receiving end somehow?

Reply:
The receiving end would have to be an actual parachain that's onboarded etc. Not sure if that's something that we want to add at this point.

I think that for the purpose of most integration tests, this setup is fine and we can leave it as it is. For other tests, such as XCM ones, we might need to use a fudge companion with multiple paras. So, I wouldn't mingle with this, for now.

Reply:
Ok Cool!

BTW, I think for the `FudgeEnv` is quite easy to add new parachains, just adding it in the `fudge::companion` struct. Something to think about in the future, but I think it's feasible.